File: musesp_ui/components/button.py
"""Button —— 按钮组件。"""

from collections.abc import Callable

import pygame

from musesp_ui.components.core import Component, Constraintable, Direction
from musesp_ui.components.label import Label
from musesp_ui.font import get_font

_DUMMY_EVENT = pygame.event.Event(pygame.USEREVENT)


class Button(Component):
    """文本按钮。

    内部包含一个居中的 Label，背景随 hover/pressed 状态变化，
    disabled 时绘制半透明黑色遮罩并忽略所有鼠标事件。

    支持 enable/disable 状态及对应事件。
    """

    def __init__(self, text: str = "", x: int = 0, y: int = 0, width: int = 120,
                 height: int = 40, font_size: int = 24):
        super().__init__(x=x, y=y, width=width, height=height)
        self.text = text
        self.font_size = font_size
        self.enabled = True
        self._handlers["enable"] = []
        self._handlers["disable"] = []
        self._overlay: pygame.Surface | None = None
        self.layout_direction = Direction.VERTICAL
        font = get_font(font_size)
        tw, th = font.size(text)
        self._label = Label(text, width=tw, height=th, font_size=font_size)
        self._label.v_constraint = Constraintable.MINIMUM
        self._label.h_constraint = Constraintable.NONE
        self._label.min_height = th
        self.add_sub_component(self._label)

    # ---- enable / disable ----

    def enable(self) -> None:
        """启用按钮，触发 ``on_enable`` 事件。"""
        if not self.enabled:
            self.enabled = True
            self._emit("enable", _DUMMY_EVENT)

    def disable(self) -> None:
        """禁用按钮，清除 hover/pressed，触发 ``on_disable`` 事件。"""
        if self.enabled:
            self.enabled = False
            self._hovered = False
            self._pressed = False
            self._emit("disable", _DUMMY_EVENT)

    def bind_on_enable(self, handler: Callable) -> None:
        """绑定启用事件。``handler(event) -> bool``。"""
        self._bind("enable", handler)

    def bind_on_disable(self, handler: Callable) -> None:
        """绑定禁用事件。``handler(event) -> bool``。"""
        self._bind("disable", handler)

    # ---- 绘制 / 事件 ----

    def _draw_self(self, surface: pygame.Surface, draw_x: int, draw_y: int) -> None:
        rect = pygame.Rect(draw_x, draw_y, self.width, self.height)
        if not self.enabled:
            bg_color = (80, 80, 80)
        elif self._pressed:
            bg_color = (100, 100, 100)
        elif self._hovered:
            bg_color = (140, 140, 140)
        else:
            bg_color = (80, 80, 80)
        pygame.draw.rect(surface, bg_color, rect)
        if not self.enabled:
            self._ensure_overlay()
            surface.blit(self._overlay, (draw_x, draw_y))

    def _ensure_overlay(self) -> None:
        if (self._overlay is not None
                and self._overlay.get_size() == (self.width, self.height)):
            return
        overlay = pygame.Surface((self.width, self.height), pygame.SRCALPHA)
        overlay.fill((0, 0, 0, 128))
        self._overlay = overlay

    def _handle_event(self, event: pygame.event.Event) -> bool:
        if not self.enabled:
            return True
        return super()._handle_event(event)

File: musesp_ui/src/components/button.rs
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;

use winit::event::WindowEvent;

use crate::components::core::{ComponentBase, ComponentTrait, Constraintable, Direction, EventHandler};
use crate::components::label::Label;
use crate::renderer::UIRenderer;

pub struct Button {
    pub base: ComponentBase,
    pub text: String,
    pub font_size: u32,
    enabled: Arc<AtomicBool>,
}

impl Button {
    pub fn new(
        text: &str,
        x: i32,
        y: i32,
        width: i32,
        height: i32,
        font_size: u32,
    ) -> Box<Self> {
        let mut base = ComponentBase::new(x, y, width, height);
        base.layout_direction = Direction::Vertical;
        let label = Label::new(text, 0, 0, 0, 0, font_size, (255, 255, 255));
        let child: Box<dyn ComponentTrait> = label;
        base.children.push(child);
        if let Some(c) = base.children.last_mut() {
            c.base_mut().h_constraint = Constraintable::None;
            c.base_mut().v_constraint = Constraintable::Minimum;
        }

        Box::new(Button {
            base,
            text: text.to_string(),
            font_size,
            enabled: Arc::new(AtomicBool::new(true)),
        })
    }

    pub fn enable(&mut self) {
        if !self.enabled.load(Ordering::Relaxed) {
            self.enabled.store(true, Ordering::Relaxed);
            self.base.emit("enable", None);
        }
    }

    pub fn disable(&mut self) {
        if self.enabled.load(Ordering::Relaxed) {
            self.enabled.store(false, Ordering::Relaxed);
            self.base.hovered = false;
            self.base.pressed = false;
            self.base.emit("disable", None);
        }
    }

    pub fn bind_enable(&mut self, handler: EventHandler) {
        self.base.bind_event("enable", handler);
    }

    pub fn bind_disable(&mut self, handler: EventHandler) {
        self.base.bind_event("disable", handler);
    }
}

impl ComponentTrait for Button {
    fn base(&self) -> &ComponentBase {
        &self.base
    }
    fn base_mut(&mut self) -> &mut ComponentBase {
        &mut self.base
    }
    fn draw_self(&self, renderer: &mut UIRenderer, dx: i32, dy: i32) {
        let bg = if !self.enabled.load(Ordering::Relaxed) {
            (80, 80, 80, 255)
        } else if self.base.pressed() {
            (100, 100, 100, 255)
        } else if self.base.hovered() {
            (140, 140, 140, 255)
        } else {
            (80, 80, 80, 255)
        };
        renderer.draw_rect(dx, dy, self.base.width, self.base.height, bg);
        if !self.enabled.load(Ordering::Relaxed) {
            renderer.draw_rect(
                dx,
                dy,
                self.base.width,
                self.base.height,
                (0, 0, 0, 128),
            );
        }
    }

    fn dispatch_event(&mut self, event: &WindowEvent) -> bool {
        if !self.enabled.load(Ordering::Relaxed) {
            return true;
        }
        self.base.dispatch_event(event)
    }
}
